### ISSUE TITLE
iscan fails for data in resource hierarchies

### ISSUE THREAD

**Issue by kript** [irods/irods#3419]
Hi folks,

We have a situation where the user that runs iRODS on the iCAT is different to the one that runs it on the iRES, at least on many of our servers.  Unfortunately, due to #3418 this means I can't run iscan on the file unless I want to fiddle with permissions, which I don't;

```
irods@irods-seq-i21:~$ iscan /irods-seq-i21-bc/fluidigm/e4/18/0f/1662062078/6e/18/ce/S059_1662062078.csv 
/irods-seq-i21-bc/fluidigm/e4/18/0f/1662062078/6e/18/ce/S059_1662062078.csv is not registered in iRODS

irods@irods-seq-i21:~$ ils -A /seq/fluidigm/e4/18/0f/1662062078/6e/18/ce/S059_1662062078.csv
  /seq/fluidigm/e4/18/0f/1662062078/6e/18/ce/S059_1662062078.csv
        ACL - ss_4112#seq:read object   trace#Sanger1:read object   dnap_ro#seq:read object   srpipe#Sanger1:own   rodsBoot#seq:own   ds12#Sanger1:read object   hb3#Sanger1:read object   irods-g1#seq:own   psdpipe#Sanger1:read object   kdj#Sanger1:read object 
```

Since both accounts have `rodsadmin` privilege I should be able to use the 'admin' flag as in other `icommands` to grant access, I suggest.

**Comment by carsten-jahn** [comment]
We have the same issue here (the workaround is to modify the file permissions, which is not great).
So, +1

**Comment by kript** [comment]
Its still an issue for us.

The situation arises with us because the IES runs as one account (say irods-seq) and the iRES have another (say irods-seq-ires). Both are rodsadmins  Users have their own accounts, which are rodsuser.

If I put a file in, as a rodsuser account, and give 'irods-seq' read access, but not 'irods-seq-ires', then when I run `iscan` on  one of the iRES (as 'irods-seq-ires' I get the aforementioned "is not registered in iRODS" error. 

As system administrators, we need to be able to run iscan as a rodsadmin account on an iRES and have it correctly confirm the presence (or lack of) a file in the resource, without replying on the fact that every file in that resource has read permissions granted to the account the `iscan` is being run as.

Does that help?

Chers

John

**Comment by trel** [comment]
Alan's work above was on a single machine - and makes sense so far (it aligned with our hoped/expected behavior).  It is captured in the table below.

| case | iscan user role | iscan user permission | unix permission | result |
| :---: | :------------: | :----------------: | :--------------------------: | :----:  |
1 | rodsadmin | own | readable | silent |
2 | rodsadmin | read | readable | silent |
3| rodsadmin | none | readable | silent |
4| rodsuser | own | readable | silent |
5 | rodsuser | read | readable | silent |
6 | rodsuser | none | readable | "is not registered in iRODS" |
7 | rodsadmin | own | none | `SYS_NO_PATH_PERMISSION` |
8 | rodsadmin | read | none | `SYS_NO_PATH_PERMISSION` |
9| rodsadmin | none | none | `SYS_NO_PATH_PERMISSION` |
10| rodsuser | own | none | `SYS_NO_PATH_PERMISSION` |
11 | rodsuser | read | none | `SYS_NO_PATH_PERMISSION` |
12 | rodsuser | none | none | `SYS_NO_PATH_PERMISSION` |

If I understand correctly, for case 3 above, you're seeing a result of "is not registered in iRODS" rather than "silent".

We'll test on two machines in the same Zone this week to try and replicate what you're seeing.

**Comment by trel** [comment]
We reproduced the setup (two iRODS 4.1.9 servers (one iCAT and one resource) in the same Zone) with two `rodsadmin` and one `rodsuser` accounts.

The `rodsuser` account put a file into the resource on the resource server.  No permissions were additionally granted.

The `rodsadmin` account (with no iRODS permissions on the data object) on the resource server ran `iscan` on the physical path of the newly put data object - and there was no output (as we see in the above case 3).

Are there any other variables we have not taken into consideration?

We have not seen the behavior you're describing.

**Comment by jaking92** [comment]
I was able to reproduce the issue by putting a file (as `rodsuser` named `fred`) into a resource and then moving the resource into a hierarchy. Run as `rodsadmin` user (no iRODS permissions on testfile.txt)...
```
$ ils -AL /tempZone/home/fred
/tempZone/home/fred:
        ACL - fred#tempZone:own   
        Inheritance - Disabled
  fred              0 storage2            3 2018-01-16.14:52 & test.txt
        generic    /tmp/testVault2/home/fred/test.txt
        ACL - fred#tempZone:own
$ iscan /tmp/testVault2/home/fred/test.txt
$ iadmin addchildtoresc testrepl storage2
$ ils -AL /tempZone/home/fred
/tempZone/home/fred:
        ACL - fred#tempZone:own   
        Inheritance - Disabled
fred              0 testroot;testrepl;storage2            3 2018-01-16.14:52 & test.txt
        generic    /tmp/testVault2/home/fred/test.txt
        ACL - fred#tempZone:own
$ iscan /tmp/testVault2/home/fred/test.txt
/tmp/testVault2/home/fred/test.txt is not registered in iRODS
```
Moving the resource back out of the hierarchy seems to make `iscan` work again. There seems to be an unintended interaction between the physical path and logical path in iRODS.

Thanks for the help, all! I will begin working on the admin flag.

**Comment by jaking92** [comment]
Changing title to reflect cause of the bug. See also: #3081

**Comment by jaking92** [comment]
After looking at this again, it appears that this issue has been resolved in 4.2.x. We will still try to work on back-porting a solution, but this is here to document how this is resolved in 4.2.x.

Below is some output demonstrating the behavior of `iscan` on a file in a resource hierarchy:
4-1-stable:
```
$ ils -L
/tempZone/home/rods:
  rods              0 pt0;ufs          490 2018-01-18.15:07 & VERSION.json
        generic    /tmp/ufsVault/home/rods/VERSION.json
$ iscan /tmp/ufsVault/home/rods/VERSION.json
/tmp/ufsVault/home/rods/VERSION.json is not registered in iRODS
$ iquest "select RESC_NAME where DATA_PATH = '/tmp/ufsVault/home/rods/VERSION.json'"
RESC_NAME = pt0
------------------------------------------------------------
```
4-2-stable:
```
$ ils -L
/tempZone/home/rods:
  rods              0 pt0;ufs1          224 2018-01-18.15:23 & VERSION.json
        generic    /tmp/ufs1Vault/home/rods/VERSION.json
$ iscan /tmp/ufs1Vault/home/rods/VERSION.json
$ iquest "select RESC_NAME where DATA_PATH = '/tmp/ufs1Vault/home/rods/VERSION.json'"
RESC_NAME = ufs1
------------------------------------------------------------
```

`iscan` operates by creating a GenQuery to see whether the file (or files) is registered with the database; it looks something like this:
```
select DATA_ID where DATA_PATH = <path-to-file> and ( RESC_LOC like <gethostname> or RESC_LOC = <gethostname> )
```
`RESC_LOC` is set to `EMPTY_RESC_HOST` for passthru resources. In 4.1.x, since the resource associated with that `DATA_PATH` is a passthru, the hostname will not match the query, and `iscan` fails. In 4.2.x, the resource associated with that `DATA_PATH` is the resource which actually contains the file, so the hostname is correct.
